native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/QueryProcessor.java

```diff
@@ -124,6 +126,83 @@ public static BStream nativeQuery(
         }
     }
 
+    public static Object nativeQueryRow(
+            BObject client, Object paramSQLString,
+            BTypedesc ballerinaType,
+            AbstractStatementParameterProcessor statementParameterProcessor,
+            AbstractResultParameterProcessor resultParameterProcessor) {
+        Object dbClient = client.getNativeData(Constants.DATABASE_CLIENT);
+        TransactionResourceManager trxResourceManager = TransactionResourceManager.getInstance();
+        if (dbClient != null) {
+            SQLDatasource sqlDatasource = (SQLDatasource) dbClient;
+            if (!((Boolean) client.getNativeData(Constants.DATABASE_CLIENT_ACTIVE_STATUS))) {
+                return ErrorGenerator.getSQLApplicationError("SQL Client is already closed, hence " +
+                        "further operations are not allowed");
+            }
+            Connection connection = null;
+            PreparedStatement statement = null;
+            ResultSet resultSet = null;
+            String sqlQuery = null;
+            try {
+                if (paramSQLString instanceof BString) {
+                    sqlQuery = ((BString) paramSQLString).getValue();
+                } else {
+                    sqlQuery = Utils.getSqlQuery((BObject) paramSQLString);
+                }
+                connection = SQLDatasource.getConnection(trxResourceManager, client, sqlDatasource);
+                statement = connection.prepareStatement(sqlQuery);
+                if (paramSQLString instanceof BObject) {
+                    statementParameterProcessor.setParams(connection, statement, (BObject) paramSQLString);
+                }
+                resultSet = statement.executeQuery();
+                if (!resultSet.next()) {
+                    throw new ApplicationError("Query retrieved an empty result.");
+                }
+
+                Type describingType = ballerinaType.getDescribingType();
+
+                // If the return data type is a record
+                if (describingType.getTag() == TypeTags.RECORD_TYPE_TAG) {
+                    RecordType recordConstraint = (RecordType) describingType;
+                    List<ColumnDefinition> columnDefinitions = Utils.getColumnDefinitions(resultSet, recordConstraint);
+                    return resultParameterProcessor.createRecord(resultSet, columnDefinitions, recordConstraint);
+                }
+
+                // If the return data type is anything other than a record
+                ColumnDefinition columnDefinition = Utils.getColumnDefinition(resultSet, 1, describingType);
+                Object retVal = resultParameterProcessor.createValue(resultSet, 1, columnDefinition);
+
+                if (resultSet.getMetaData().getColumnCount() > 1) {
+                    throw new ApplicationError("Query retrieved more than one column.");
+                }
```

Comment:
Do we need to send an ApplicationError here? since the remote method name is `queryRow` the ballerina user may expect to get the first row of the query.

Reply:
This error is only thrown if the expected return type is **not** a record. (e.g. when the user runs a query like `COUNT(*)`).  The discussion with @daneshk was to return an error if the query retrieved more than one column or more than one row.

If the expected return type is a record, there will be no error generated. Only the first row is returned.